File: android-app/core-database/src/main/java/com/chainlesschain/android/core/database/util/SqlLike.kt
package com.chainlesschain.android.core.database.util

/**
 * SQLite LIKE-pattern escaping (Android shared helper).
 *
 * Escape the LIKE metacharacters (% _ \) so a user-supplied value is matched
 * LITERALLY. Room binds `:query` params, so this is not an injection vector —
 * but an unescaped % or _ in the value still acts as a wildcard and broadens a
 * search to unrelated rows (and, for an id/prefix match, could hit the WRONG
 * one). Mirrors packages/cli/src/lib/sql-like.js and
 * packages/personal-data-hub/lib/sql-like.js.
 *
 * Room usage: the escaped value only takes effect when the `@Query` pairs the
 * LIKE with `ESCAPE '\'`. Keep the DAO's public search method wrapping the raw
 * `@Query` method so callers pass plain text and escaping is centralized, e.g.
 *
 *     @Query("... WHERE name LIKE '%' || :q || '%' ESCAPE '\\' ...")
 *     fun searchRaw(q: String): Flow<List<T>>
 *
 *     fun search(query: String): Flow<List<T>> = searchRaw(SqlLike.escapeLike(query))
 */
object SqlLike {
    private val META = Regex("""([\\%_])""")

    /** Escape % _ \ in [value] for a literal LIKE match (pair with ESCAPE '\'). */
    fun escapeLike(value: String?): String =
        (value ?: "").replace(META, """\\$1""")

    /** `<escaped>%` — a literal-prefix LIKE pattern (use with ESCAPE '\'). */
    fun likePrefix(value: String?): String = escapeLike(value) + "%"

    /** `%<escaped>%` — a literal-substring LIKE pattern (use with ESCAPE '\'). */
    fun likeContains(value: String?): String = "%" + escapeLike(value) + "%"
}

File: packages/personal-data-hub/lib/sql-like.js
/**
 * SQLite LIKE-pattern escaping (PDH shared helper).
 *
 * Escape the LIKE metacharacters (% _ \) so a user-supplied value is matched
 * LITERALLY in a `... LIKE ? ESCAPE '\'` clause. Without escaping, a value
 * containing % or _ acts as a wildcard — broadening a search to unrelated rows
 * (and, for an id/prefix match that resolves or mutates a record, hitting the
 * WRONG one). Mirrors packages/cli/src/lib/sql-like.js.
 *
 * Always pair the returned pattern with `ESCAPE '\'` in the SQL. For an
 * anchored (prefix) match, keep a non-empty guard at the call site — an empty
 * value yields a bare `%`, which still matches all rows.
 */

/** Escape % _ \ in `value` for a literal LIKE match. */
function escapeLike(value) {
  return String(value ?? "").replace(/[\\%_]/g, "\\$&");
}

/** `<escaped value>%` — a literal-prefix LIKE pattern (use with ESCAPE '\'). */
function likePrefix(value) {
  return escapeLike(value) + "%";
}

/** `%<escaped value>%` — a literal-substring LIKE pattern (use with ESCAPE '\'). */
function likeContains(value) {
  return "%" + escapeLike(value) + "%";
}

module.exports = { escapeLike, likePrefix, likeContains };
